In [1]:
pip install notebook

Público objetivo 󰚒
📌Premisa
Hoy día, los datos se han convertido en el mayor activo que poseen las
organizaciones para fundamentar sus decisiones y dirigir estrategias
empresariales efectivas; esto va desde poder analizar patrones de
comportamiento hasta prever tendencias futuras.
En este contexto, entra InnovateTech, una empresa líder en tecnología que, con
nuestra ayuda, busca construir y mantener su plataforma de datos. Esta
plataforma no solo respalda la toma de decisiones estratégicas a través del
análisis e interpretación de datos, sino que también permite que la empresa pueda
examinar y organizar su información con diferentes propósitos: ajustar sus
productos y servicios, detectar nichos de mercado, expandirse, mejorar su atención
al cliente, entre otros.
En este caso, el desafío al que se enfrentan es comprender mejor a su público
objetivo. Están buscando, con nuestro apoyo, entender a fondo a sus
consumidores: a qué se dedican, qué hacen, qué edad tienen, qué les gusta, qué les
duele, y qué necesitan. Esto les permitirá optimizar, contextualizar y personalizar
los mensajes hacia sus clientes lo que a su vez les ayudará a definir mejores
campañas y obtener leads de mayor calidad, es decir, potenciales clientes más
alineados con sus productos y servicios.
Como miembro del equipo de Ingeniería de Datos de la empresa y aplicando los
conocimientos adquiridos en la unidad de Excel, Python y análisis de datos, te
enfocarás en extraer, analizar y gestionar datos clave.
🎯 Propósito general
Manipulación de Datos con Python:
● Utiliza la biblioteca Pandas en Python para cargar un archivo CSV que
contiene datos de ventas de diferentes sucursales de la empresa.
● Limpia y prepara los datos utilizando funciones de manipulación de datos en
Python, como la eliminación de filas duplicadas y el manejo de valores nulos.

2

Análisis de Datos con Python:
● Utiliza NumPy para realizar cálculos de álgebra lineal, como la multiplicación
de matrices y el cálculo de determinantes.
● Utiliza Pandas para generar un reporte descriptivo que incluya estadísticas
descriptivas básicas y tablas dinámicas sobre las ventas de las diferentes
sucursales.
Presentación de Informes con Excel:
● Exporta los datos limpios y el reporte descriptivo generado en Python a un
archivo Excel.
● Utiliza fórmulas y funciones de Excel para calcular métricas adicionales,
como el total de ventas y el promedio de ventas por sucursal.
● Aplica formatos personalizados y estilos visuales para mejorar la
presentación de la información en el archivo Excel.
📭Entregables
➢ Archivo Excel que contiene los datos limpios, el reporte descriptivo y las
métricas adicionales calculadas.
➢ Archivo .py con el código utilizado para el análisis de datos en Python.
📝¿Qué vamos a validar?
Estos son algunos puntos que vamos a validar:
A. Cumplimiento de lo requerido por archivo:
✓ Utiliza Pandas: para cargar y limpiar los datos de ventas.
✓ Utiliza NumPy: para realizar cálculos de álgebra lineal según sea
necesario.
✓ Utiliza Pandas: para generar el reporte descriptivo de ventas.
✓ Exporta los datos y el reporte a un archivo Excel.

3

✓ Utiliza Excel: para calcular métricas adicionales y mejorar la
presentación del informe.

B. Calidad del código: Programación en Python
✓ Modularización: Revisar si el código está organizado en módulos o
funciones de manera lógica, facilitando su comprensión,
mantenimiento y reutilización.
✓ Documentación: Comprobar que el código esté adecuadamente
documentado, incluyendo comentarios explicativos, descripciones de
funciones y notas sobre la lógica utilizada.
✓ Buenas prácticas: Evaluar si se siguen las buenas prácticas de
programación, como nombramiento coherente de variables,
estructuras de código limpias, y uso eficiente de recursos.

C. Eficiencia y optimización:
✓ Rendimiento del Código: Revisar el código en busca de posibles
ineficiencias que puedan ralentizar la ejecución o consumir recursos
innecesarios.
✓ Optimización de Consultas: Asegurar que las consultas SQL y las
operaciones de Pandas se realicen de manera eficiente,
especialmente en el manejo de grandes volúmenes de datos.

D. Seguridad y privacidad:
✓ Manejo de Datos Sensibles: Verificar que se respeten las prácticas de
seguridad y privacidad de datos, especialmente si se manejan datos
personales o confidenciales.
✓ Protección contra Vulnerabilidades: Asegurarse de que el código esté
libre de vulnerabilidades comunes, como inyecciones SQL en las
consultas.

E. Presentación y reporte de resultados:
✓ Claridad en los Informes: Comprobar que los informes generados sean
claros, bien estructurados y que presenten los hallazgos de manera
comprensible.

4

✓ Fundamentación de Conclusiones: Asegurarse de que las
conclusiones y recomendaciones estén bien fundamentadas y
basadas en los datos y análisis realizados.

📄Requerimientos
→ Analizar los archivos proporcionados para comprender su estructura y los
datos que contienen🗄 :
■ ecommerce_customers_dataset.csv
■ ecommerce_orders_dataset.csv
■ ecommerce_order_items_dataset.csv
■ ecommerce_products_dataset.csv
■ ecommerce_order_payments_dataset.csv
→ Realizar las siguientes actividades con pandas 🐼:
■ Cargar los archivos CSV en DataFrames de Pandas
■ Establecer la columna índices de las tablas como la clave primaria en
el DataFrame, ejemplo para customers la clave será customer_id.
■ Obtener el número total de clientes únicos en el conjunto de datos
■ Calcular el promedio de valor de pago por pedido
■ Determinar la categoría de producto más vendida
■ Calcular el número total de pedidos realizadoslds in line 3, saw 1290

procedemos a la instalación e importacion de librerias

In [1]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [4]:
import pandas as pd

In [5]:
import os

In [6]:
df_orders = pd.read_csv("https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_orders_dataset.csv")
df_orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [7]:
df_order_items = pd.read_csv("https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_items_dataset.csv")
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [8]:
df_products = pd.read_csv("https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_products_dataset.csv")
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [9]:
df_customers = pd.read_csv("https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_customers_dataset.csv")
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [10]:
df_order_payments = pd.read_csv("https://raw.githubusercontent.com/Andresmup/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_payments_dataset.csv")
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [11]:
# defino funcion para saber como se compone cada uno de los dataFrame a analizar
def datosDataFrame(df):
    print("--------------------------------------------------")
    print("Tamaño del DataFrame:")
    print(df.shape)
    print("--------------------------------------------------")
    # Guardo el resultado de df.info() en una variable antes de imprimirlo, para evitar que imprima None.
    print("Análisis de tipo de datos:")
    df_info = df.info()
    print("--------------------------------------------------")
    print("Recuento de nulos:")
    print(df.isnull().sum())
    print("--------------------------------------------------")
    print("Recuento de duplicados:")
    print(df.duplicated().sum())
    print("--------------------------------------------------")


una vez realizada mi funcion comienzo a ver como se componen cada uno de mis DataFrame

In [12]:
datosDataFrame(df_orders)

--------------------------------------------------
Tamaño del DataFrame:
(99441, 8)
--------------------------------------------------
Análisis de tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
--------------------------------------------------
Recuento de nulos:
order_id                      

In [13]:
datosDataFrame(df_order_items)

--------------------------------------------------
Tamaño del DataFrame:
(112650, 7)
--------------------------------------------------
Análisis de tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
--------------------------------------------------
Recuento de nulos:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price      

In [14]:
datosDataFrame(df_products)

--------------------------------------------------
Tamaño del DataFrame:
(32951, 9)
--------------------------------------------------
Análisis de tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
--------------------------------------------------


In [15]:
datosDataFrame(df_customers)

--------------------------------------------------
Tamaño del DataFrame:
(99441, 5)
--------------------------------------------------
Análisis de tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
--------------------------------------------------
Recuento de nulos:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
--------------------------------------------------
Recuento de duplic

In [16]:
datosDataFrame(df_order_payments)

--------------------------------------------------
Tamaño del DataFrame:
(103886, 5)
--------------------------------------------------
Análisis de tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
--------------------------------------------------
Recuento de nulos:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
--------------------------------------------------
Recuento de duplicados:
0
-----------

TRANSFORMACION DATAFRAMES
Para asegurar que nuestro analisis no tenga problemas a futuro es importante transformar los tipos de datos, ademas de asegurarse que los datos sean homogeneos.

En aquellos dataframe que hay nulos si la proporcion de nulos no es muy grande es factible eliminar dichas filas.

Funcion transformarFechas

In [17]:
#Definimos nuestra función transformarFechas
def transformarFechas(df, columnas_fechas): #Los parametros son el dataframe y la lista con las columnas a transformar
    df[columnas_fechas] = df[columnas_fechas].apply(pd.to_datetime) #Transforma las columnas de columnas_fechas a datetime
    return df
     

Funcion transformarString

In [18]:
#Definimos nuestra función transformarString
def transformarString(df, columnas_string): #Los parametros son el dataframe y la lista con las columnas a transformar
    for columna in columnas_string: #Transforma las columnas en columnas_string a string
        df[columna] = df[columna].astype("string")
    return df

Funcion convertirMayusculas

In [19]:
#Definimos nuestra función convertirMayusculas
def convertirMayusculas(df, columna): #Los parametros son el dataframe y la lista con las columna a transformar
    df[columna] = df[columna].str.upper() # Convertir todos los valores de la columna a mayusculas
    return df

Funcion convertirMinusculas

In [20]:
#Definimos nuestra funcion convertirMinusculas
def convertirMinusculas(df, columna): #Los parametros son el dataframe y la lista con las columna a transformar
    df[columna] = df[columna].str.lower() # Convertir todos los valores de la columna a minúsculas
    return df

Orders

In [21]:
# Porcentaje de nulos en cada columna
null_percentages_orders = (df_orders.isnull().sum() / len(df_orders)) * 100

#Veamos cual es el % de nulos
print(null_percentages_orders)
     

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64


In [22]:
# Filtramos las columnas con menos del 10% de valores nulos
columns_to_keep_orders = null_percentages_orders[null_percentages_orders <= 10].index

# Eliminamos las filas con nulos solo en las columnas seleccionadas
df_orders.dropna(subset=columns_to_keep_orders, inplace=True)

#Chequeo cantidad de null
df_orders.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [23]:
# Lista de columnas que contienen fechas
columnas_fechas_orders = ['order_purchase_timestamp', 'order_approved_at',
                          'order_delivered_carrier_date', 'order_delivered_customer_date',
                          'order_estimated_delivery_date']

#Aplicamos nuestra funcion conversora de fechas
df_orders = transformarFechas(df_orders, columnas_fechas_orders)

In [24]:
# Lista de columnas a transfomar en string
columnas_string_orders = ["order_id","customer_id","order_status"]

#Aplicamos nuestra funcion conversora de string
df_orders = transformarString(df_orders, columnas_string_orders)

In [25]:
#Revisamos que los tipos de datos sean correctos
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  string        
 1   customer_id                    96461 non-null  string        
 2   order_status                   96461 non-null  string        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96461 non-null  datetime64[ns]
 6   order_delivered_customer_date  96461 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96461 non-null  datetime64[ns]
dtypes: datetime64[ns](5), string(3)
memory usage: 6.6 MB


In [26]:
#Aplicamos la conversion a minusculas
df_orders = convertirMinusculas(df_orders, "order_status")

Order items

In [27]:
# Lista de columnas que contienen fechas
columnas_fechas_order_items = ['shipping_limit_date']

#Aplicamos nuestra funcion conversora de fechas
df_order_items = transformarFechas(df_order_items, columnas_fechas_order_items)

In [28]:
# Lista de columnas a transfomar en string
columnas_string_order_items = ["order_id","order_item_id","product_id","seller_id"]

#Aplicamos nuestra funcion conversora de string
df_order_items = transformarString(df_order_items, columnas_string_order_items)

Products

In [29]:
# Porcentaje de nulos en cada columna
null_percentages_products = (df_products.isnull().sum() / len(df_products)) * 100

print(null_percentages_products)

product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64


In [30]:
# Filtramos las columnas con menos del 10% de valores nulos
columns_to_keep_products = null_percentages_products[null_percentages_products <= 10].index

# Eliminamos las filas con nulos solo en las columnas seleccionadas
df_products.dropna(subset=columns_to_keep_products, inplace=True)

#Chequeo cantidad de null
df_products.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [31]:
# Lista de columnas a transfomar en string
columnas_string_products = ["product_id","product_category_name"]

#Aplicamos nuestra funcion conversora de string
df_products = transformarString(df_products, columnas_string_products)

In [32]:
#Revisamos que los tipos de datos sean correctos
df_products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  string 
 1   product_category_name       32340 non-null  string 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), string(2)
memory usage: 2.5 MB


In [33]:
#Aplicamos la conversion a minusculas
df_products = convertirMinusculas(df_products, "product_category_name")

Customers

In [34]:
# Lista de columnas a transfomar en string
columnas_string_customers = ["customer_id","customer_unique_id","customer_zip_code_prefix","customer_city","customer_state"]

#Aplicamos nuestra funcion conversora de string
df_customers = transformarString(df_customers, columnas_string_customers)

In [35]:
#Revisamos que los tipos de datos sean correctos
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB


In [36]:
#Aplicamos la conversion a minusculas
df_customers = convertirMinusculas(df_customers, "customer_city")

In [37]:
#Aplicamos la conversion a mayusculas
df_customers = convertirMayusculas(df_customers, "customer_state")

Payments

In [38]:
# Lista de columnas a transfomar en string
columnas_string_order_payments = ["order_id","payment_type"]

#Aplicamos nuestra funcion conversora de string
df_order_payments = transformarString(df_order_payments, columnas_string_order_payments)

In [39]:
#Aplicamos la conversion a minusculas
df_order_payments = convertirMinusculas(df_order_payments, "payment_type")

ANALISIS DATAFRAMES

In [40]:
#Definimos una función para obtener insights relevantes de nuestros dataframes
def analisisDataFrames(df): #Como parametro tenemos los dataframes
  print("----------")
  print("Análisis variables cuantitativas")
  columnas_numericas = df.select_dtypes(include=['int', 'float']).columns #Extraemos las columnas numericas
  if not columnas_numericas.empty: #Aplicamos un if en caso de que no tenga columnas numéricas el dataframe
    print(df[columnas_numericas].describe()) #Usamos el describe para aquellas columnas cuantitativas
  print("----------")
  print("----------")
  print("Análisis variables cualitativas")
  columnas_string = df.select_dtypes(include=['string']).columns #Extraemos las columnas string
  if not columnas_string.empty: #Aplicamos un if en caso de que no tenga columnas del tipo string el dataframe
    print(df[columnas_string].describe()) #Usamos el describe para aquellas columnas cualitativas

veamos que info nos arroja el df orders si aplicamos la funcion analisis_dataframe

Análisis de tendencias de ventas a lo largo del tiempo

In [41]:

# Ordenar el DataFrame por 'order_purchase_timestamp'
df_orders = df_orders.sort_values(by='order_purchase_timestamp')

# Calcular las tendencias de ventas mensuales
monthly_sales = df_orders.set_index('order_purchase_timestamp').resample('M').size()

print("Tendencias de ventas mensuales:")
print(monthly_sales)

Tendencias de ventas mensuales:
order_purchase_timestamp
2016-09-30       1
2016-10-31     270
2016-11-30       0
2016-12-31       1
2017-01-31     748
2017-02-28    1641
2017-03-31    2546
2017-04-30    2303
2017-05-31    3545
2017-06-30    3135
2017-07-31    3872
2017-08-31    4193
2017-09-30    4149
2017-10-31    4478
2017-11-30    7288
2017-12-31    5513
2018-01-31    7069
2018-02-28    6556
2018-03-31    7003
2018-04-30    6798
2018-05-31    6749
2018-06-30    6096
2018-07-31    6156
2018-08-31    6351
Freq: M, dtype: int64


 Se observa cierta estacionalidad en las ventas, con picos en ciertos meses y valles en otros. Por ejemplo, parece haber un aumento en las ventas hacia finales de año, con picos notables en noviembre y diciembre. Esto sugiere que hay períodos específicos en los que la demanda de tus productos o servicios es más alta, lo que podría estar relacionado con días festivos, temporadas de regalos o promociones especiales.

In [42]:
# Agrupar las ventas por mes y calcular las ventas totales
monthly_sales = df_orders.set_index('order_purchase_timestamp').resample('M').size()

# Encontrar el mes con el mayor número de ventas
max_month_sales = monthly_sales.idxmax()

print("El mes que más se vende en este período es:", max_month_sales.strftime('%B'))

El mes que más se vende en este período es: November


In [43]:
# que valor tiene el producto que mas se vende en Noviembre
november_orders = df_orders[(df_orders['order_purchase_timestamp'].dt.month == 11)]

# Fusionar los pedidos de noviembre con los detalles de los productos
november_orders_details = pd.merge(november_orders, df_order_items, on='order_id')

# Calcular las ventas totales de cada producto en noviembre
product_sales_november = november_orders_details.groupby('price')['price'].sum()

# Encontrar el producto más vendido en noviembre
top_product_november = product_sales_november.idxmax()

print("El producto más vendido en noviembre tiene un valor de :", top_product_november,"reales")

El producto más vendido en noviembre tiene un valor de : 59.9 reales


El hecho de que un producto con un precio más bajo sea el más vendido en el mes con mayores ventas se puede deber a una demanda particularmente fuerte para productos más accesibles durante noviembre. Esto podría ser útil para ajustar estrategias de precios o promociones para aprovechar esta demanda.

In [44]:
#  cantidad de productos por categoría
category_distribution = df_products['product_category_name'].value_counts()

print("Distribución de productos por categoría:")
print(category_distribution)

Distribución de productos por categoría:
product_category_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: count, Length: 73, dtype: Int64



"La distribución de productos por categoría muestra una clara concentración en ciertas categorías, como 'cama_mesa_banho', 'esporte_lazer', 'moveis_decoracao', 'beleza_saude' y 'utilidades_domesticas', que representan un alto volumen de productos vendidos. Esto sugiere una fuerte demanda por productos relacionados con el hogar, el ocio, la decoración y el cuidado personal. Por otro lado, existen categorías con una menor representación en las ventas, como 'fashion_roupa_infanto_juvenil', 'casa_conforto_2', 'pc_gamer', 'seguros_e_servicos' y 'cds_dvds_musicais', que podrían ser nichos de mercado menos explorados o productos menos populares entre los clientes. Este insight destaca la importancia de comprender las preferencias del consumidor y ajustar la estrategia de ventas en consecuencia, centrándose en las categorías con mayor demanda para maximizar las oportunidades de venta y satisfacer las necesidades del mercado."

In [45]:

# Agrupar los datos por estado y contar el número de clientes únicos en cada estado
state_customers_count = df_customers.groupby('customer_state')['customer_unique_id'].nunique()

# Ordenar los resultados en orden ascendente
state_customers_count_sorted = state_customers_count.sort_values()

# Obtener los estados que menos compran (los primeros 5 estados con menos clientes únicos)
states_with_least_customers = state_customers_count_sorted.head(5)

print("Los estados que menos compran son:")
print(states_with_least_customers)

Los estados que menos compran son:
customer_state
RR     45
AP     67
AC     77
AM    143
RO    240
Name: customer_unique_id, dtype: int64


In [46]:
# Agrupar los datos por estado y contar el número de clientes únicos en cada estado
state_customers_count = df_customers.groupby('customer_state')['customer_unique_id'].nunique()

# Ordenar los resultados en orden descendente
state_customers_count_sorted = state_customers_count.sort_values(ascending=False)

# Obtener los estados que más compran (los primeros 5 estados con más clientes únicos)
states_with_most_customers = state_customers_count_sorted.head(5)

print("Los estados que más compran son:")
print(states_with_most_customers)

Los estados que más compran son:
customer_state
SP    40302
RJ    12384
MG    11259
RS     5277
PR     4882
Name: customer_unique_id, dtype: int64



Estos informes proporcionan una visión interesante de la distribución geográfica de las ventas. Algunas observaciones sobre los estados que menos y más compran:

Estados que menos compran:
Los estados de RR (Roraima), AP (Amapá), AC (Acre), AM (Amazonas) y RO (Rondônia) muestran un menor número de compras en comparación con otros estados. Esto podría estar relacionado con factores como la densidad de población, el acceso a servicios de entrega y la actividad económica en esas regiones.
Estados que más compran:
Por otro lado, los estados de SP (São Paulo), RJ (Río de Janeiro), MG (Minas Gerais), RS (Río Grande do Sul) y PR (Paraná) muestran una mayor actividad de compra. Estos estados son conocidos por tener una mayor densidad de población, economías más desarrolladas y una infraestructura de comercio electrónico más establecida, lo que puede contribuir a un mayor número de compras.
Estos informes podrían ser útiles para identificar áreas geográficas donde se pueden enfocar los esfuerzos de marketing y ventas para mejorar las ventas en los estados con menor actividad comercial, mientras se mantienen estrategias efectivas para satisfacer la demanda en los estados con mayor actividad come

En resumen, la mayor cantidad de ventas se concentra en la región sureste de Brasil, mientras que la menor cantidad de ventas se observa en la región norte del país. Estas tendencias pueden influir en las estrategias de mercado y ventas para optimizar el rendimiento en cada región y satisfacer las necesidades únicas de los clientes en cada área.rcial.

In [47]:
datosDataFrame(df_order_payments)

--------------------------------------------------
Tamaño del DataFrame:
(103886, 5)
--------------------------------------------------
Análisis de tipo de datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  string 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  string 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), string(2)
memory usage: 4.0 MB
--------------------------------------------------
Recuento de nulos:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
--------------------------------------------------
Recuento de duplicados:
0
------------

In [48]:
payment_distribution = df_order_payments['payment_type'].value_counts(normalize=True)

# Explorar la cantidad total de pagos realizados con cada método de pago
total_payment_value = df_order_payments.groupby('payment_type')['payment_value'].sum()

print("Distribución de métodos de pago:")
print(payment_distribution)
print("\nCantidad total de pagos realizados con cada método de pago:")
print(total_payment_value)

Distribución de métodos de pago:
payment_type
credit_card    0.739224
boleto          0.19044
voucher         0.05559
debit_card     0.014718
not_defined    0.000029
Name: proportion, dtype: Float64

Cantidad total de pagos realizados con cada método de pago:
payment_type
boleto          2869361.27
credit_card    12542084.19
debit_card       217989.79
not_defined           0.00
voucher          379436.87
Name: payment_value, dtype: float64


In [49]:
# Filtrar los pagos realizados con vouchers
voucher_payments = df_order_payments[df_order_payments['payment_type'] == 'voucher']

# Calcular el valor total de los pagos realizados con vouchers
total_voucher_payments = voucher_payments['payment_value'].sum()

# Calcular el valor total de todas las ventas
total_sales = df_order_payments['payment_value'].sum()

# Calcular el porcentaje que representan los vouchers con respecto a las ventas totales
percentage_of_sales = (total_voucher_payments / total_sales) * 100

print("Los vouchers representan aproximadamente el", round(percentage_of_sales, 2), "% de las ventas totales.")

Los vouchers representan aproximadamente el 2.37 % de las ventas totales.


In [50]:
# Filtrar los pagos realizados con boleto bancario
boleto_payments = df_order_payments[df_order_payments['payment_type'] == 'boleto']

# Calcular el valor total de los pagos realizados con boleto bancario
total_boleto_payments = boleto_payments['payment_value'].sum()

# Calcular el porcentaje que representan los boletos bancarios con respecto a las ventas totales
percentage_of_boleto_sales = (total_boleto_payments / total_sales) * 100

print("Los boletos bancarios representan aproximadamente el", round(percentage_of_boleto_sales, 2), "% de las ventas totales.")


Los boletos bancarios representan aproximadamente el 17.92 % de las ventas totales.


In [51]:
# Filtrar los pagos realizados con tarjeta de crédito
credit_card_payments = df_order_payments[df_order_payments['payment_type'] == 'credit_card']

# Calcular el valor total de los pagos realizados con tarjeta de crédito
total_credit_card_payments = credit_card_payments['payment_value'].sum()

# Calcular el porcentaje que representan las tarjetas de crédito con respecto a las ventas totales
percentage_of_credit_card_sales = (total_credit_card_payments / total_sales) * 100

print("Las tarjetas de crédito representan aproximadamente el", round(percentage_of_credit_card_sales, 2), "% de las ventas totales.")


Las tarjetas de crédito representan aproximadamente el 78.34 % de las ventas totales.


In [52]:
# Filtrar los pagos realizados con tarjeta de débito
debit_card_payments = df_order_payments[df_order_payments['payment_type'] == 'debit_card']

# Calcular el valor total de los pagos realizados con tarjeta de débito
total_debit_card_payments = debit_card_payments['payment_value'].sum()

# Calcular el porcentaje que representan las tarjetas de débito con respecto a las ventas totales
percentage_of_debit_card_sales = (total_debit_card_payments / total_sales) * 100

print("Las tarjetas de débito representan aproximadamente el", round(percentage_of_debit_card_sales, 2), "% de las ventas totales.")


Las tarjetas de débito representan aproximadamente el 1.36 % de las ventas totales.


"Según los datos analizados, se observa una clara preferencia por el uso de tarjetas de crédito como método de pago en comparación con otros métodos disponibles. Las tarjetas de crédito representan la mayor parte de las ventas totales, aproximadamente el 78.34%, lo que sugiere una fuerte confianza del consumidor en este método de pago. Por otro lado, aunque las tarjetas de débito también son una opción común, representan solo el 1.36% de las ventas totales. Los boletos bancarios, aunque menos populares que las tarjetas de crédito, aún representan una parte significativa de las ventas totales, aproximadamente el 17.92%. Los vouchers, por su parte, representan la menor proporción de las ventas totales, con aproximadamente el 2.37%. Estos hallazgos destacan la importancia de ofrecer una variedad de métodos de pago para satisfacer las preferencias de los clientes y maximizar las oportunidades de venta."





In [53]:
# Contamos cuántas veces aparece cada cliente en el DataFrame de pedidos
order_count = df_customers['customer_unique_id'].value_counts()

# Contamos cuántos clientes aparecen solo una vez y cuántos más de una vez
one_time_customers = (order_count == 1).sum()  # Clientes que compran solo una vez
repeat_customers = (order_count > 1).sum()  # Clientes que compran más de una vez

# Calculamos los porcentajes
total_customers = len(df_customers)
percentage_one_time = (one_time_customers / total_customers) * 100
percentage_repeat = (repeat_customers / total_customers) * 100

print("Porcentaje de clientes que compran solo una vez:", percentage_one_time)
print("Porcentaje de clientes que compran más de una vez:", percentage_repeat)

Porcentaje de clientes que compran solo una vez: 93.62234893052161
Porcentaje de clientes que compran más de una vez: 3.0138474070051586


con respecto a este analisis sobre la retención de Clientes, Es fundamental entender por qué la mayoría de los clientes solo hacen una compra. Realizar encuestas de satisfacción, recopilar comentarios y analizar el comportamiento de compra podrían proporcionar información valiosa sobre cómo retener a los clientes y convertirlos en compradores recurrentes.
En resumen, estos datos indican una oportunidad para mejorar la retención y fidelización de los clientes, lo que podría conducir a un aumento en las ventas y la rentabilidad a largo plazo.

In [3]:
# Crear la carpeta 'data' si no existe
if not os.path.exists('data'):
    os.makedirs('data')

In [54]:
df_orders.to_csv("data/df_orders_limpio.csv")

In [55]:
df_order_items.to_csv("data/df_order_items_limpio.csv")

In [56]:
df_products.to_csv("data/df_products_limpio.csv")

In [57]:
df_customers.to_csv("data/df_customers_limpio.csv")

In [58]:
df_order_payments.to_csv("data/df_order_payments_limpio.csv")